### PR TITLE
clang-tidy: readability-redundant-smartptr-get does not remove (#97964)

### PR DESCRIPTION
**Pull request by akshaykumars614**
added a check to remove '->' if exists

added testcase and modified Release Notes

pull request in continue of https://github.com/llvm/llvm-project/pull/98757

### CODE REVIEW

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 1]
```diff
@@ -498,6 +498,10 @@ Changes in existing checks
   false-positives when type of the member does not match the type of the
   initializer.
 
+- Improved :doc:`readability-redundant-smartptr-get
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:501]: you may need to rebase, as release notes has been cleared after branch out

**Review by PiotrZSL — COMMENTED** [review]
Except release notes, looks fine.

File: clang-tools-extra/docs/ReleaseNotes.rst (hunk 1 above)
Comment (akshaykumars614) [reply clang-tools-extra/docs/ReleaseNotes.rst:501]: will deleting the old branch work?

Comment (akshaykumars614) [reply clang-tools-extra/docs/ReleaseNotes.rst:501]: I cannot see those markers in my local branch

**Comment by akshaykumars614** [comment]
I made the changes

File: clang-tools-extra/test/clang-tidy/checkers/readability/redundant-smartptr-get.cpp [hunk 2]
```diff
@@ -20,6 +20,47 @@ struct shared_ptr {
...
+template <typename T>
+struct vector {
+  vector();
+  bool operator==(const vector<T>& other) const;
+  bool operator!=(const vector<T>& other) const;
+  unsigned long size() const;
+  bool empty() const;
+
+  // Basic iterator implementation for testing
+  struct iterator {
```
Comment (njames93) [inline clang-tools-extra/test/clang-tidy/checkers/readability/redundant-smartptr-get.cpp:32]: This boilerplate is excessive, you can just say the iterator is a pointer to T.
Also definitions aren't needed for all methods in the class, a declaration is more than enough

**Review by njames93 — CHANGES_REQUESTED** [review]

File: clang-tools-extra/test/clang-tidy/checkers/readability/redundant-smartptr-get.cpp (hunk 2 above)
Comment (akshaykumars614) [reply clang-tools-extra/test/clang-tidy/checkers/readability/redundant-smartptr-get.cpp]: modified

Comment (akshaykumars614) [reply clang-tools-extra/test/clang-tidy/checkers/readability/redundant-smartptr-get.cpp]: Got it. Made the changes.

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 3]
```diff
@@ -104,6 +104,296 @@ New check aliases
 Changes in existing checks
 ^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+- Improved :doc:`bugprone-assert-side-effect
+  <clang-tidy/checks/bugprone/assert-side-effect>` check by detecting side
+  effect from calling a method with non-const reference parameters.
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:109]: still incorrectly rebased code, that stuff shoudn't be here
just rebase your changes on top of main branch, and then do force push, or cleanup this fill manually

Comment (akshaykumars614) [reply clang-tools-extra/docs/ReleaseNotes.rst]: got it. Changed accordingly.

Comment (akshaykumars614) [reply clang-tools-extra/docs/ReleaseNotes.rst]: Did I do mistake again...? I am really sorry if I did. I am new here. I am learning online and doing it. Please correct me @PiotrZSL

File: clang-tools-extra/docs/ReleaseNotes.rst [hunk 4]
```diff
@@ -104,6 +104,10 @@ New check aliases
 Changes in existing checks
 ^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+- Improved :doc:`readability-redundant-smartptr-get
+  <clang-tidy/checks/readability/redundant-smartptr-get>` check to
+  remove '->', when reduntant get() is removed.
```
Comment (PiotrZSL) [inline clang-tools-extra/docs/ReleaseNotes.rst:109]:
```suggestion
  remove `->`, when reduntant `get()` is removed.
```

**Review by PiotrZSL — APPROVED** [review]

**Pull request merged**